### PR TITLE
feat(cstor-upgrade): add upgrade runtask for cStor  0.8.2 to 0.9.0

### PR DESCRIPTION
**Pull request by sonasingh46**
This PR does following :
1. Add runtasks for cStor pool upgrade to version 0.9.0 from 0.8.2
2. [WIP] Add runtasks for cStor volume upgrade to version 0.9.0 from 0.8.2

Signed-off-by: Ashutosh Kumar <ashutosh.kumar@openebs.io>

### CODE REVIEW

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 1]
```diff
@@ -0,0 +1,775 @@
...
+# CASTemplate cstor-volume-update-082-090 is
+# used to upgrade a cstor volume
+apiVersion: openebs.io/v1alpha1
+kind: CASTemplate
+metadata:
+  name: cstor-volume-update-082-090
+spec:
+  defaultConfig:
+  - name: baseversion
+    value: "0.8.0"
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Change it to 0.8.2

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 2]
```diff
@@ -0,0 +1,775 @@
...
+
+# CASTemplate cstor-volume-update-082-090 is
+# used to upgrade a cstor volume
+apiVersion: openebs.io/v1alpha1
+kind: CASTemplate
+metadata:
+  name: cstor-volume-update-082-090
+spec:
+  defaultConfig:
+  - name: baseversion
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Do a camel case naimg

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 3]
```diff
@@ -0,0 +1,775 @@
...
+    value: "0.8.0"
+  - name: targetversion
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Came case naming

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 4]
```diff
@@ -0,0 +1,775 @@
...
+#     - upgrade-cstor-volume-082-090-post-check-patch-deployment-image
+#     - upgrade-cstor-volume-082-090-patch-target-svc
+#     - upgrade-cstor-volume-082-090-post-check-target-svc
+#     - upgrade-cstor-volume-082-090-patch-cstor-volume-cr
+#     - upgrade-cstor-volume-082-090-post-check-cstor-volume-cr
+#     - upgrade-cstor-volume-082-090-patch-cstor-volume-replica
+#     - upgrade-cstor-volume-082-090-post-check-cstor-volume-replicas
+#     - upgrade-cstor-volume-082-090-list-target-replicaset
+#     - upgrade-cstor-volume-082-090-list-target-pod
+#     - upgrade-cstor-volume-082-090-delete-stale-replicaset-target
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Put comments for runtask -- describing what it does.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 5]
```diff
@@ -0,0 +1,775 @@
...
+spec:
+  meta: |
+    id: patchResult
+    apiVersion: openebs.io/v1alpha1
+    kind: UpgradeResult
+    action: patch
+    objectName: {{ .UpgradeItem.upgradeResultName }}
+    runNamespace: default
+  post: |
+    {{ $retries := 0 -}}
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Not required

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml [hunk 6]
```diff
@@ -0,0 +1,775 @@
...
+  meta: |
+    id: getvoldetails
+    apiVersion: v1
+    kind: PersistentVolume
+    action: get
+    objectName: {{ .UpgradeItem.name }}
+    runNamespace: {{ .UpgradeItem.namespace }}
+  post: |
+    {{- jsonpath .JsonResult "{.metadata.annotations.openebs\\.io/cas-type}" | trim | saveAs "getvoldetails.volCASType" .TaskResult | noop -}}
+    {{- .TaskResult.getvoldetails.volCASType | notFoundErr "volume CAS type not found" | saveIf "getvoldetails.notFoundErr" .TaskResult | noop -}}
```
Comment (sonasingh46) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-volume-update-082-090.yaml]: Came case `getvoldetails`

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+    - upgrade-cstor-pool-082-090-patch-upgrade-result
+    # patches pool deployment with target version
+    - upgrade-cstor-pool-082-090-patch-deployment-image
+    # Waits until new pods have came up successfully
+    - upgrade-cstor-pool-082-090-patch-deployment-image-status
+    # checks whether pool containers in target version or not
+    - upgrade-cstor-pool-082-090-post-check-patch-deployment-image
+    # lists all replicasets of pool deployment
+    - upgrade-cstor-pool-082-090-patch-sp-version
+    - upgrade-cstor-pool-082-090-patch-sp-version-post-check
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: do we want to provide comments here as well?

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 8]
```diff
@@ -0,0 +1,664 @@
...
+    - upgrade-cstor-pool-082-090-patch-csp-version-post-check
+    - upgrade-cstor-pool-082-090-patch-deployment-version
+    - upgrade-cstor-pool-082-090-patch-deployment-version-post-check
+    - upgrade-cstor-pool-082-090-list-replicaset
+    # list working replicaset and pod of pool deployment
+    - upgrade-cstor-pool-082-090-list-pod
+    - upgrade-cstor-pool-082-090-delete-replicaset
+  taskNamespace: default
+---
+# upgrade-cstor-pool-082-090-get-cstorpool gets details of CStorPool CR.
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]:
```yaml
# upgrade-cstor-pool-082-090-get-cstorpool fetches the details of 
# CStorPool CR. These details are used in other runtask(s) later.
# This should always run.
```

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 9]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $status =.Config.successStatus.value }}
+    {{- else }}
+    {{- $status =.Config.failStatus.value }}
+    {{- end }}
+
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-get-cstorpool" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: This is a limitation.
Job may not be running in default namespace.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 10]
```diff
@@ -0,0 +1,664 @@
...
+    apiVersion: openebs.io/v1alpha1
+    kind: CStorPool
+    action: get
+    objectName: {{ .UpgradeItem.name }}
+    disable: false
+  post: |
+    {{- jsonpath .JsonResult "{.metadata.uid}" | trim | saveAs "CStorPool.uid" .TaskResult | noop -}}
+    {{- jsonpath .JsonResult "{.metadata.labels.openebs\\.io/version}" | trim | saveAs "CStorPool.labels" .TaskResult | noop -}}
+
+    {{- $message := printf "Successfully got details of CStorPool {%s}." .UpgradeItem.name -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: $message should also participate in below if else logic

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 11]
```diff
@@ -0,0 +1,664 @@
...
+
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-get-cstorpool" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    {{- upgradeResultUpdateTasks $taskStartTime $URName $URNamespace $taskName $taskStatus $taskMessage $taskEndTime -}}
+    {{- if eq $status .Config.failStatus.value }}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Can have a linespace before trying the if condition?

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 12]
```diff
@@ -0,0 +1,664 @@
...
+    {{- verifyErr "CStorPool version is not 0.8.2" true | saveAs "CStorPool.verifyErr" .TaskResult | noop -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Is it possible to have all the message defined in post at one place?
We can provide extra info like CSP name etc in these messages
e.g.
```yaml
{{- $successMsg := "" -}}
{{- $errorMsg := "" -}}
{{- $verifyErrMsg := "" -}}
```

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 7 above)
Comment (mittachaitu) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: IMO RunTask itself describe what it does

Comment (AmitKumarDas) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Lets follow one standard. Either have them or dont have them if runtask names are sufficient.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 13]
```diff
@@ -0,0 +1,664 @@
...
+    apiVersion: extensions/v1beta1
+    kind: Deployment
+    action: get
+    objectName: {{ .UpgradeItem.name }}
+    runNamespace: {{ .UpgradeItem.namespace }}
+    disable: false
+  post: |
+    {{- jsonpath .JsonResult "{.spec.template.spec.containers[?(@.name=='cstor-pool')].image}" | trim | saveAs "poolDeployment.cstorpoolimage" .TaskResult | noop -}}
+    {{- jsonpath .JsonResult "{.spec.template.spec.containers[?(@.name=='cstor-pool-mgmt')].image}" | trim | saveAs "poolDeployment.cstorpoolmgmtimage" .TaskResult | noop -}}
+    {{- jsonpath .JsonResult "{.metadata.labels.openebs\\.io/version}" | trim | saveAs "poolDeployment.labels" .TaskResult | noop -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: should we save it as .poolDeployment.version

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 14]
```diff
@@ -0,0 +1,664 @@
...
+
+    {{- $message := printf "Successfully got details of CStorPool deployment {%s}." .UpgradeItem.name -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: same as earlier comment. $message should participate in if else block

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 15]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-get-deployment" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    {{- upgradeResultUpdateTasks $taskStartTime $URName $URNamespace $taskName $taskStatus $taskMessage $taskEndTime -}}
+    {{- if eq $status .Config.failStatus.value }}
+    {{- verifyErr "CStorPool deployment version is not 0.8.2" true | saveAs "poolDeployment.verifyErr" .TaskResult | noop -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: check my earlier comments on grouping all messages together & here just use the variable e.g. `$errMsgInvalidVersion`
```yaml
{{- verifyErr $errMsgInvalidVersion true | saveAs "poolDeployment.verifyErr" .TaskResult | noop -}}
```

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 16]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    {{- upgradeResultUpdateTasks $taskStartTime $URName $URNamespace $taskName $taskStatus $taskMessage $taskEndTime -}}
+    {{- if eq $status .Config.failStatus.value }}
+    {{- verifyErr "CStorPool deployment version is not 0.8.2" true | saveAs "poolDeployment.verifyErr" .TaskResult | noop -}} 
+    {{- end }}
+---
+# upgrade-cstor-pool-082-090-get-cstorpool gets details of StoragePool CR.
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Check my earlier comments where I have tried to rephrase

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 17]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-get-storagepool" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    {{- upgradeResultUpdateTasks $taskStartTime $URName $URNamespace $taskName $taskStatus $taskMessage $taskEndTime -}}
+    {{- if eq $status .Config.failStatus.value }}
+    {{- verifyErr "StoragePool version is not 0.8.2" true | saveAs "StoragePool.verifyErr" .TaskResult | noop -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: why do we throw verify error if subsequent tasks can disable based on their own version

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 12 above)
Comment (sonasingh46) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Once place we can do -- if required at multiple places. IMO, it just fine if we are passing at one place.
We will see later if required.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 11 above)
Comment (sonasingh46) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Will fix it.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 9 above)
Comment (sonasingh46) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: cc @imazik @sagarkrsd

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 10 above)
Comment (sonasingh46) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: will fix it.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 18]
```diff
@@ -0,0 +1,664 @@
...
+  name: upgrade-cstor-pool-082-090-get-storagepool
+  namespace: default
+spec:
+  meta: |
+    id: StoragePool
+    apiVersion: openebs.io/v1alpha1
+    kind: StoragePool
+    action: get
+    objectName: {{ .UpgradeItem.name }}
+    disable: false
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: no need to mention this as by default this is false

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 19]
```diff
@@ -0,0 +1,664 @@
...
+    {{ $isOldCStorPool := contains .Config.baseversion.value .TaskResult.poolDeployment.cstorpoolimage }}
+    {{ $isOldCStorPoolMGMT := contains .Config.baseversion.value .TaskResult.poolDeployment.cstorpoolmgmtimage  }}
+    {{ $shouldpatch := or $isOldCStorPool $isOldCStorPoolMGMT | toString }}
+    id: patchDeploymentImage
+    apiVersion: extensions/v1beta1
+    kind: Deployment
+    action: patch
+    objectName: {{ .UpgradeItem.name }}
+    runNamespace: {{ .UpgradeItem.namespace }}
+    disable: {{ eq $shouldpatch "false" }}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: if this check is present, then why do we throw verify errors in other runtasks

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 20]
```diff
@@ -0,0 +1,664 @@
...
+      spec:
+        template:
+          metadata:
+            labels:
+              openebs.io/version: {{ .Config.targetversion.value }}
+              openebs.io/cstor-pool: {{ .UpgradeItem.name }} 
+          spec:
+            containers:
+            - name: cstor-pool
+              image: quay.io/openebs/cstor-pool:v0.9.x-ci
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: can we remove this hardcoding & instead make use of cas config or runtime config

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 21]
```diff
@@ -0,0 +1,664 @@
...
+                  - "/bin/sh"
+                  - "-c"
+                  - zfs set io.openebs:livenesstimestap='$(date)' cstor-$OPENEBS_IO_CSTOR_ID
+                failureThreshold: 3
+                initialDelaySeconds: 300
+                periodSeconds: 10
+                successThreshold: 1
+                timeoutSeconds: 30
+            - name: cstor-pool-mgmt
+              image: quay.io/openebs/cstor-pool-mgmt:v0.9.x-ci
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: same comment

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 22]
```diff
@@ -0,0 +1,664 @@
...
+              # Setting ports to nil, since it is not required and
+              # 9500 port is used by m-exporter
+              ports:
+            - name: maya-exporter
+              image: quay.io/openebs/m-exporter:v0.9.x-ci
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: same comment

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 12 above)
Comment (sonasingh46) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Also -- it has some format specifier. Not sure how will it behave.
Will take it later.

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml (hunk 17 above)
Comment (AmitKumarDas) [reply k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: If we throw verify error how do we handle retrying an upgrade that had failed earlier

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 23]
```diff
@@ -0,0 +1,664 @@
...
+apiVersion: openebs.io/v1alpha1
+kind: RunTask
+metadata:
+  name: upgrade-cstor-pool-082-090-patch-deployment-image-status
+  namespace: default
+spec:
+  meta: |
+    {{ $isOldCStorPool := contains .TaskResult.poolDeployment.cstorpoolimage .Config.baseversion.value }}
+    {{ $isOldCStorPoolMGMT := eq .TaskResult.poolDeployment.cstorpoolmgmtimage "" }}
+    {{ $shouldpatch := or $isOldCStorPool $isOldCStorPoolMGMT | toString }}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: are we using this?

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 24]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $message := printf "%s name: {%s} namespace: {%s}" .TaskResult.patchDeploymentImageStatus.rolloutStatus .UpgradeItem.name .UpgradeItem.namespace -}}
+
+
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-patch-deployment-image-status" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: Ideal case, only endtime should get updated. if starttime is not set already then reuse the endtime. But we should not override the starttime

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 25]
```diff
@@ -0,0 +1,664 @@
...
+    apiVersion: extensions/v1beta1
+    kind: Deployment
+    action: rolloutstatus
+    objectName: {{ .UpgradeItem.name }}
+    runNamespace: {{ .UpgradeItem.namespace }}
+    retry: "20,20s"
+  post: |
+    {{- jsonpath .JsonResult "{.isRolledout}" | trim | saveAs "patchDeploymentImageStatus.isRolledout" .TaskResult | noop -}}
+    {{- jsonpath .JsonResult "{.message}" | trim | saveAs "patchDeploymentImageStatus.rolloutStatus" .TaskResult | noop -}}
+    {{- $status := "" -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: let us have a linespace

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 26]
```diff
@@ -0,0 +1,664 @@
...
+
+    {{- $taskStatus := upgradeResultWithTaskStatus $status -}}
+    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
+    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-pool-082-090-patch-deployment-image-status" -}}
+    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
+    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
+    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
+    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
+    {{- if eq $status .Config.failStatus.value }}
+    {{- verifyErr "Pool deployment roll out not successfull" true | saveAs "patchDeploymentImageStatus.verifyErr" .TaskResult | noop -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: "Failed to rollout pool deployment"

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 27]
```diff
@@ -0,0 +1,664 @@
...
+    {{ $isNewCStorPoolMGMT := contains .Config.targetversion.value .TaskResult.postCheckDeploymentImagePatch.cstorpoolmgmtimage }}
+    {{ $isNewMayaExporter := contains .Config.targetversion.value .TaskResult.postCheckDeploymentImagePatch.mayaexporterimage }}
+
+    {{- $status := "" -}}
+    {{- if and $isNewCStorPool $isNewCStorPoolMGMT $isNewMayaExporter }}
+    {{- $status = .Config.successStatus.value -}}
+    {{- else }}
+    {{- $status = .Config.failStatus.value -}}
+    {{- end }}
+    {{- $taskName := "upgrade-cstor-pool-082-090-post-check-patch-deployment-image" -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: let us have a linespace before $taskName

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 28]
```diff
@@ -0,0 +1,664 @@
...
+    {{- $message := printf "pool image :{%s} pool mgmt image :{%s} maya-exporter image : {%s}" .TaskResult.postCheckDeploymentImagePatch.cstorpoolimage .TaskResult.postCheckDeploymentImagePatch.cstorpoolmgmtimage .TaskResult.postCheckDeploymentImagePatch.mayaexporterimage -}}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: we are missing some logic probably?
Where is $status & $message used?

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 29]
```diff
@@ -0,0 +1,664 @@
...
+apiVersion: openebs.io/v1alpha1
+kind: RunTask
+metadata:
+  name: upgrade-cstor-pool-082-090-post-check-patch-deployment-image
+  namespace: default
+spec:
+  meta: |
+    {{ $isOldCStorPool := contains .TaskResult.poolDeployment.cstorpoolimage .Config.baseversion.value }}
+    {{ $isOldCStorPoolMGMT := eq .TaskResult.poolDeployment.cstorpoolmgmtimage "" }}
+    {{ $shouldpatch := or $isOldCStorPool $isOldCStorPoolMGMT | toString }}
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: can we rename to $isOldVersion

File: k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml [hunk 30]
```diff
@@ -0,0 +1,750 @@
...
+spec:
+  meta: |
+    id: podList
+    apiVersion: v1
+    kind: Pod
+    action: list
+    runNamespace: {{ .UpgradeItem.namespace }}
+    options: |-
+      labelSelector: app=cstor-pool,openebs.io/storage-pool-claim=cstor-sparse-pool
+    disable: false
```
Comment (AmitKumarDas) [inline k8s/upgrades/0.8.2-0.9.0/cstor/cstor-pool-update-082-090.yaml]: let us remove this as this is the default

**Pull request merged**